packages/contracts/contracts/protocol/oracles/BalancerOracle.sol

```diff
@@ -57,124 +63,189 @@ library BalancerOracle {
 		return price; 
 			
 	}
-		//calc metastable using getRate(); 	//calc composable stable using getRate(); 
 	
-
 	/// @dev Return fair reserve amounts given spot reserves, weights, and fair prices.
-  	/// @param resA Reserve of the first asset
-  	/// @param resB Reserve of the second asset
-  	/// @param wA Weight of the first asset
-  	/// @param wB Weight of the second asset
-  	/// @param pxA Fair price of the first asset
-  	/// @param pxB Fair price of the second asset
-  	function computeFairReserves(
-  	  uint resA,
-  	  uint resB,
-  	  uint wA,
-  	  uint wB,
-  	  uint pxA,
-  	  uint pxB
-  	) internal pure returns (uint fairResA, uint fairResB) {
-  	  // NOTE: wA + wB = 1 (normalize weights)
-  	  // constant product = resA^wA * resB^wB
-  	  // constraints:
-  	  // - fairResA^wA * fairResB^wB = constant product
-  	  // - fairResA * pxA / wA = fairResB * pxB / wB
-  	  // Solving equations:
-  	  // --> fairResA^wA * (fairResA * (pxA * wB) / (wA * pxB))^wB = constant product
-  	  // --> fairResA / r1^wB = constant product
-  	  // --> fairResA = resA^wA * resB^wB * r1^wB
-  	  // --> fairResA = resA * (resB/resA)^wB * r1^wB = resA * (r1/r0)^wB
-  	  uint r0 = BNum.bdiv(resA, resB);
-  	  uint r1 = BNum.bdiv(BNum.bmul(wA, pxB), BNum.bmul(wB, pxA));
-  	  // fairResA = resA * (r1 / r0) ^ wB
-  	  // fairResB = resB * (r0 / r1) ^ wA
-	  
-	  uint256 ratio; 
-  	  if (r0 > r1) {
-		ratio = BNum.bdiv(r1, r0);
-  	    fairResA = BNum.bmul(resA, BNum.bpow(ratio, wB));
-  	    fairResB = BNum.bdiv(resB, BNum.bpow(ratio, wA));
-  	  } else {
-  	    ratio = BNum.bdiv(r0, r1);
-  	    fairResA = BNum.bdiv(resA, BNum.bpow(ratio, wB));
-  	    fairResB = BNum.bmul(resB, BNum.bpow(ratio, wA));
-  	  }
-	  //check for low pool ratios
-	  require(ratio > 1e12, "reserves too low"); 
-  	}
+	/// @param reserves Reserves of the assets
+	/// @param weights Weights of the assets
+	/// @param prices Fair prices of the assets
+	function computeFairReserves(
```

Comment:
I dont understand what is happening here so I'll just comment on the general code guidelines:

switch all for loops to

`for (uint256 i; i < length;) {
unchecked{++i}
}`

cache lengths: uint256 length = reserves.length

also all of these variables could be initialized before first for loop, and then reused in every iteration

Reply:
I copied these formulas directly from midas. I think if we want to avoid another audit, we should keep the code exactly the same as the audited code from midas. Maybe we can include these in the next audit?

Reply:
Ok, sounds good. These nested for loops look suspicious though.
We should still include more tests in edge cases where pools become unbalanced

Reply:
I removed this